crates/storage/nippy-jar/src/compression/zstd.rs

```diff
@@ -0,0 +1,175 @@
+use crate::{compression::Compression, NippyJarError};
+use serde::{Deserialize, Serialize};
+use std::{
+    fs::File,
+    io::{Read, Write},
+};
+use zstd::bulk::{Compressor, Decompressor};
+
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub enum ZstdState {
+    #[default]
+    PendingDictionary,
+    Ready,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+/// Zstd compression structure. Supports a compression dictionary per column.
+pub struct Zstd {
+    /// State. Should be ready before compressing.
+    pub(crate) state: ZstdState,
+    /// Uses custom dictionaries to compress data.
+    pub(crate) use_dict: bool,
+    /// Max size of a dictionary
+    pub(crate) max_dict_size: usize,
+    /// List of column dictionaries.
+    pub(crate) raw_dictionaries: Option<Vec<Vec<u8>>>,
+    /// Number of columns to compress.
+    columns: usize,
+}
+
+impl Zstd {
+    /// Creates new [`Zstd`].
+    pub fn new(use_dict: bool, max_dict_size: usize, columns: usize) -> Self {
+        // TODO add level
+        Self {
+            state: if use_dict { ZstdState::PendingDictionary } else { ZstdState::Ready },
+            use_dict,
+            max_dict_size,
+            raw_dictionaries: None,
+            columns,
+        }
+    }
+
+    /// If using dictionaries, creates a list of [`Decompressor`].
+    pub fn generate_decompressors(&self) -> Result<Vec<Decompressor<'_>>, NippyJarError> {
+        if let Some(dictionaries) = &self.raw_dictionaries {
+            return Ok((0..self.columns)
+                .map(|column| Decompressor::with_dictionary(&dictionaries[column]))
+                .collect::<Result<Vec<_>, _>>()?)
+        }
+        Ok(vec![])
+    }
+
+    /// If using dictionaries, creates a list of [`Compressor`].
+    pub fn generate_compressors(&self) -> Result<Option<Vec<Compressor>>, NippyJarError> {
+        match self.state {
+            ZstdState::PendingDictionary => Err(NippyJarError::CompressorNotReady),
+            ZstdState::Ready => {
+                if !self.use_dict {
+                    return Ok(None)
+                }
+
+                let mut compressors = None;
+                if let Some(dictionaries) = &self.raw_dictionaries {
+                    let mut cmp = Vec::with_capacity(dictionaries.len());
+
+                    for dict in dictionaries {
+                        cmp.push(Compressor::with_dictionary(0, dict)?);
+                    }
+                    compressors = Some(cmp)
+                }
+                Ok(compressors)
+            }
+        }
+    }
+
+    /// Compresses a value using a dictionary.
+    pub fn compress_with_dictionary(
+        &self,
+        value: &[u8],
+        tmp_buf: &mut Vec<u8>,
```

Comment:
do we need to expose this or it can be just a temporary buffer length?
```suggestion
        buf_len: usize,
```

Reply:
i'm not sure what are you suggesting with `buf_len`, but `tmp_buf` is there to avoid reallocations when compressing many values

Reply:
that makes sense, I suggest passing the expected temporary buffer length instead of a mutable reference to a temporary buffer, because iiuc we don't really need to use a temporary buffer from the caller side.

Reply:
but the goal is for the caller not having to reallocate new buffers every time it is called